### PR TITLE
Threat map optimization

### PR DESCRIPTION
**Pull request by navnit-crest**
## Proposed commit message

Fixes for the Threat Map:

- **Fix invalid SVG path**: Filtered out records with same location values for source
and destination in advance.
- **Dashboard fix**: Fix search query to include CCS (Cross Cluster Search) indices by default.
- **Optimize Vega resource usage**: Vega now avoids pulling the entire _source, reducing request size and resource consumption (CPU/Memory).
- **Resolve animation field issue**: Paths are now animated by default, even if the animate field is not present. The path remains idle only when the animate key explicitly has a false value.

## Checklist

- [x] I have reviewed [tips for building integrations](https://github.com/elastic/integrations/blob/main/docs/tips_for_building_integrations.md) and this pull request is aligned with them.
- [ ] I have verified that all data streams collect metrics or logs.
- [x] I have added an entry to my package's `changelog.yml` file.
- [x] I have verified that Kibana version constraints are current according to [guidelines](https://github.com/elastic/elastic-package/blob/master/docs/howto/stack_version_support.md#when-to-update-the-condition).
- [x] I have verified that any added dashboard complies with Kibana's [Dashboard good practices](https://docs.elastic.dev/ux-guidelines/data-viz/dashboard-good-practices)  

## How to test this PR locally

1. Clone integrations repo.
2. Install elastic package locally.
3. Start elastic stack using elastic-package.
4. Move to integrations/packages/threat_map directory.
5. Run the following command to run tests. 
> elastic-package test

```
--- Test results for package: threat_map - START ---
╭────────────┬─────────────┬───────────┬─────────────────────────────────────────────────────────────────────┬────────┬──────────────╮
│ PACKAGE    │ DATA STREAM │ TEST TYPE │ TEST NAME                                                           │ RESULT │ TIME ELAPSED │
├────────────┼─────────────┼───────────┼─────────────────────────────────────────────────────────────────────┼────────┼──────────────┤
│ threat_map │             │ asset     │ dashboard threat_map-fecb4669-516d-4de9-aa98-713aca39fa6c is loaded │ PASS   │       2.44µs │
╰────────────┴─────────────┴───────────┴─────────────────────────────────────────────────────────────────────┴────────┴──────────────╯
--- Test results for package: threat_map - END   ---
Done
```

### Before
before

### After
after

### CODE REVIEW

File: packages/threat_map/kibana/dashboard/threat_map-fecb4669-516d-4de9-aa98-713aca39fa6c.json [hunk 1]
```diff
@@ -93,7 +93,7 @@
...
...e\": {\n\t\t\t\t\"enter\": {\n\t\t\t\t\t\"startAngle\": { \"value\": 0 },\n\t\t\t\t\t\"endAngle\": { \"value\": 6.29 },\n\t\t\t\t\t\"innerRadius\": { \"value\": 0 },\n\t\t\t\t\t\"stroke\": [\n\t\t\t\t\t\t{ \"test\": \"datum._source.color \u0026\u0026 datum._source.color != ''\", \"field\": \"_source.color\" },\n\t\t\t\t\t\t{\"value\": \"#54B399\"}\n\t\t\t\t\t],\n\t\t\t\t\t\"strokeWidth\": { \"value\": 2 }\n\t\t\t\t},\n\t\t\t\t\"update\": {\n\t\t\t\t\t\"x\": [\n\t\t\t\t\t\t{\"test\": \"datum._source.pulse_at_source \u0026\u0026 datum._source.pulse_at_source == true\", \"field\": \"sourceX\"},\n\t\t\t\t\t\t{\"field\": \"targetX\"}\n\t\t\t\t\t],\n\t\t\t\t\t\"y\": [\n\t\t\t\t\t\t{\"test\": \"datum._source.pulse_at_source \u0026\u0026 datum._source.pulse_at_source == true\", \"field\": \"sourceY\"},\n\t\t\t\t\t\t{\"field\": \"targetY\"}\n\t\t\t\t\t],\n\t\t\t\t\t\"outerRadius\": { \"signal\": \"radius\" },\n\t\t\t\t\t\"strokeOpacity\": { \"signal\": \"1 - fraction2\" },\n\t\t\t\t\t\"opacity\": [\n\t\t\t\t\t\t{ \"test\": \"datum._source.animate \u0026\u0026 datum._source.animate == true\", \"value\": 1 },\n\t\t\t\t\t\t{ \"value\": 0 }\n\t\t\t\t\t]\n\t\t\t\t}\n\t\t\t}\n        }\n    ]\n}"
```
Comment (chrisberkhout) [inline packages/threat_map/kibana/dashboard/threat_map-fecb4669-516d-4de9-aa98-713aca39fa6c.json:96]: Diff of the JSON in that string:

```diff
--- visspec1.json	2024-12-17 15:38:04.137814528 +0100
+++ visspec2.json	2024-12-17 15:37:56.131201312 +0100
@@ -1,600 +1,594 @@
 {
   "$schema": "https://vega.github.io/schema/vega/v5.json",
   "config": {
     "kibana": {
       "type": "map",
       "latitude": 10,
       "longitude": 10,
       "zoom": 1.3,
       "mapStyle": true,
       "zoomControl": true,
       "scrollWheelZoom": false,
       "delayRepaint": true,
       "tooltips": {
         "position": "top",
         "padding": 15,
         "textTruncate": true
       }
     }
   },
   "signals": [
     {
       "name": "zoom",
-      "description": "The current zoom level of the map",
       "value": 1
     },
     {
       "name": "latitude",
-      "description": "The starting latitude of the map",
       "value": 0
     },
     {
       "name": "longitude",
-      "description": "The starting longitude of the map",
       "value": 0
     },
     {
       "name": "width",
-      "description": "The width of the map",
       "value": 900
     },
     {
       "name": "height",
-      "description": "The height of the map",
       "value": 560
     },
     {
       "name": "timer",
       "value": 0,
       "on": [
         {
           "events": {
             "type": "timer",
             "throttle": 10
           },
-          "update": "if(timer >= 1e6, 0, timer + 10)"
+          "update": "if(timer >= 1e4, 0, timer + 10)"
         }
       ]
     },
     {
       "name": "fraction2",
-      "description": "A fraction to control the animation of the circle",
       "value": 0,
       "on": [
         {
           "events": {
             "signal": "timer"
           },
           "update": "(timer % 500) / 500"
         }
       ]
     },
     {
-      "name": "radius",
-      "description": "The animated radius of an arc",
-      "update": "20 * fraction2"
-    },
-    {
       "name": "showLables",
       "value": true,
       "bind": {
         "input": "checkbox"
       }
     }
   ],
   "data": [
     {
       "name": "routes",
       "url": {
-        "index": "_all",
+        "index": "_all,*:_all",
         "body": {
           "size": 10000,
+          "_source": false,
+          "fields": [
+            "source.geo.location.lat",
+            "source.geo.location.lon",
+            "destination.geo.location.lat",
+            "destination.geo.location.lon",
+            "source.ip",
+            "destination.ip",
+            "source.geo.country_name",
+            "destination.geo.country_name",
+            "color",
+            "weight",
+            "animate",
+            "source_label",
+            "destination_label",
+            "pulse_at_source"
+          ],
           "query": {
             "bool": {
               "must": [
                 "%dashboard_context-must_clause%",
                 {
                   "range": {
                     "@timestamp": {
                       "%timefilter%": true
                     }
                   }
                 },
                 {
                   "exists": {
                     "field": "source.geo.location"
                   }
                 },
                 {
                   "exists": {
                     "field": "destination.geo.location"
                   }
                 }
               ],
               "must_not": [
                 "%dashboard_context-must_not_clause%"
               ],
               "filter": [
                 "%dashboard_context-filter_clause%"
               ]
             }
           }
         }
       },
       "format": {
         "property": "hits.hits"
       },
       "transform": [
         {
-          "type": "filter",
-          "expr": "datum._source.source && datum._source.source.geo && datum._source.source.geo.location && datum._source.source.geo.location.lat && datum._source.source.geo.location.lon && datum._source.destination && datum._source.destination.geo && datum._source.destination.geo.location && datum._source.destination.geo.location.lat && datum._source.destination.geo.location.lon"
+          "type": "formula",
+          "expr": "isDefined(datum.fields) && isArray(datum.fields['color']) && datum.fields['color'][0] != null ? datum.fields['color'][0] : '#54B399'",
+          "as": "color"
+        },
+        {
+          "type": "formula",
+          "expr": "isDefined(datum.fields) && isArray(datum.fields['animate']) && datum.fields['animate'][0] != null && datum.fields['animate'][0] == false? 0 : 1",
+          "as": "animateOpacity"
+        },
+        {
+          "type": "formula",
+          "expr": "isDefined(datum.fields) && isArray(datum.fields['weight']) && datum.fields['weight'][0] != null? datum.fields['weight'][0] : 1",
+          "as": "weight"
+        },
+        {
+          "type": "formula",
+          "expr": "isDefined(datum.fields) && isArray(datum.fields['source_label']) && datum.fields['source_label'][0] != null? datum.fields['source_label'][0] : ''",
+          "as": "source_label"
+        },
+        {
+          "type": "formula",
+          "expr": "isDefined(datum.fields) && isArray(datum.fields['destination_label']) && datum.fields['destination_label'][0] != null? datum.fields['destination_label'][0] : ''",
+          "as": "destination_label"
+        },
+        {
+          "type": "formula",
+          "expr": "isDefined(datum.fields) && isArray(datum.fields['pulse_at_source']) && datum.fields['pulse_at_source'][0] != null? datum.fields['pulse_at_source'][0] : false",
+          "as": "pulse_at_source"
+        },
+        {
+          "type": "formula",
+          "expr": "isDefined(datum.fields) && isArray(datum.fields['source.geo.country_name']) && datum.fields['source.geo.country_name'][0] != null? datum.fields['source.geo.country_name'][0] : ''",
+          "as": "source_country"
+        },
+        {
+          "type": "formula",
+          "expr": "isDefined(datum.fields) && isArray(datum.fields['destination.geo.country_name']) && datum.fields['destination.geo.country_name'][0] != null? datum.fields['destination.geo.country_name'][0] : ''",
+          "as": "destination_country"
+        },
+        {
+          "type": "formula",
+          "expr": "isDefined(datum.fields) && isArray(datum.fields['source.ip']) && datum.fields['source.ip'][0] != null? datum.fields['source.ip'][0] : ''",
+          "as": "source_ip"
+        },
+        {
+          "type": "formula",
+          "expr": "isDefined(datum.fields) && isArray(datum.fields['destination.ip']) && datum.fields['destination.ip'][0] != null? datum.fields['destination.ip'][0] : ''",
+          "as": "destination_ip"
         },
         {
           "type": "geopoint",
           "projection": "projection",
           "fields": [
-            "_source.source.geo.location.lon",
-            "_source.source.geo.location.lat"
+            "fields['source.geo.location.lon']",
+            "fields['source.geo.location.lat']"
           ],
           "as": [
             "sourceX",
             "sourceY"
           ]
         },
         {
           "type": "geopoint",
           "projection": "projection",
           "fields": [
-            "_source.destination.geo.location.lon",
-            "_source.destination.geo.location.lat"
+            "fields['destination.geo.location.lon']",
+            "fields['destination.geo.location.lat']"
           ],
           "as": [
             "targetX",
             "targetY"
           ]
         },
         {
+          "type": "filter",
+          "expr": "datum.sourceX !== datum.targetX || datum.sourceY !== datum.targetY"
+        },
+        {
           "type": "linkpath",
           "shape": "curve",
           "sourceX": {
             "field": "sourceX"
           },
           "sourceY": {
             "field": "sourceY"
           },
           "targetX": {
             "field": "targetX"
           },
           "targetY": {
             "field": "targetY"
           }
         },
         {
           "type": "formula",
           "as": "control1X",
           "expr": "datum.sourceX + 0.2 * ((datum.targetX - datum.sourceX) + (datum.targetY - datum.sourceY))"
         },
         {
           "type": "formula",
           "as": "control1Y",
           "expr": "datum.sourceY + 0.2 * ((datum.targetY - datum.sourceY) - (datum.targetX - datum.sourceX))"
         },
         {
           "type": "formula",
           "as": "control2X",
           "expr": "datum.targetX + 0.2 * ((datum.targetY - datum.sourceY) - (datum.targetX - datum.sourceX))"
         },
         {
           "type": "formula",
           "as": "control2Y",
           "expr": "datum.targetY - 0.2 * ((datum.targetX - datum.sourceX) + (datum.targetY - datum.sourceY))"
         },
         {
           "type": "formula",
           "as": "tooltip_data",
-          "expr": "{source_ountry: datum._source.source.geo.country_name,destination_country: datum._source.destination.geo.country_name, source_IP: datum._source.source.ip, destination_IP: datum._source.destination.ip}"
+          "expr": "{source_country: datum.source_country, destination_country: datum.destination_country, source_IP: datum.source_ip, destination_IP: datum.destination_ip}"
         },
         {
           "type": "formula",
           "as": "pathLength",
           "expr": "sqrt(pow(datum.control1X - datum.sourceX, 2) + pow(datum.control1Y - datum.sourceY, 2)) + sqrt(pow(datum.control2X - datum.control1X, 2) + pow(datum.control2Y - datum.control1Y, 2)) + sqrt(pow(datum.targetX - datum.control2X, 2) + pow(datum.targetY - datum.control2Y, 2))"
         },
         {
           "type": "formula",
           "as": "duration",
           "expr": "datum.pathLength * 5"
         }
       ]
     },
     {
       "name": "animatedRoutes",
       "source": "routes",
       "transform": [
         {
           "type": "formula",
           "as": "fraction",
           "expr": "(timer % datum.duration) / datum.duration"
         }
       ]
     }
   ],
   "marks": [
     {
       "name": "linkpath-curve",
-      "description": "The curve generated using the linkpath transform",
       "type": "path",
       "from": {
         "data": "routes"
       },
       "encode": {
         "enter": {
-          "strokeWidth": [
-            {
-              "test": "datum._source.weight && datum._source.weight != ''",
-              "field": "_source.weight"
-            },
-            {
-              "value": 1
-            }
-          ],
-          "stroke": [
-            {
-              "test": "datum._source.color && datum._source.color != ''",
-              "field": "_source.color"
-            },
-            {
-              "value": "#54B399"
-            }
-          ]
+          "strokeWidth": {
+            "field": "weight"
+          },
+          "stroke": {
+            "field": "color"
+          }
         },
         "update": {
           "path": {
             "field": "path"
           }
         }
       }
     },
     {
       "name": "moving-path",
-      "description": "The stroke that moves on the curved path",
       "type": "path",
       "from": {
         "data": "animatedRoutes"
       },
       "encode": {
         "enter": {
           "strokeWidth": {
             "value": 2
           },
-          "stroke": [
-            {
-              "test": "datum._source.color && datum._source.color != ''",
-              "field": "_source.color"
-            },
-            {
-              "value": "#54B399"
-            }
-          ]
+          "stroke": {
+            "field": "color"
+          }
         },
         "update": {
           "path": {
             "signal": "'M' + (pow(1 - (datum.fraction - 0.05), 3) * datum.sourceX + 3 * pow(1 - (datum.fraction - 0.05), 2) * (datum.fraction - 0.05) * datum.control1X + 3 * (1 - (datum.fraction - 0.05)) * pow((datum.fraction - 0.05), 2) * datum.control2X + pow((datum.fraction - 0.05), 3) * datum.targetX) + ',' + (pow(1 - (datum.fraction - 0.05), 3) * datum.sourceY + 3 * pow(1 - (datum.fraction - 0.05), 2) * (datum.fraction - 0.05) * datum.control1Y + 3 * (1 - (datum.fraction - 0.05)) * pow((datum.fraction - 0.05), 2) * datum.control2Y + pow((datum.fraction - 0.05), 3) * datum.targetY) + 'L' + (pow(1 - datum.fraction, 3) * datum.sourceX + 3 * pow(1 - datum.fraction, 2) * datum.fraction * datum.control1X + 3 * (1 - datum.fraction) * pow(datum.fraction, 2) * datum.control2X + pow(datum.fraction, 3) * datum.targetX) + ',' + (pow(1 - datum.fraction, 3) * datum.sourceY + 3 * pow(1 - datum.fraction, 2) * datum.fraction * datum.control1Y + 3 * (1 - datum.fraction) * pow(datum.fraction, 2) * datum.control2Y + pow(datum.fraction, 3) * datum.targetY)"
           },
-          "opacity": [
-            {
-              "test": "datum._source.animate && datum._source.animate == true",
-              "value": 1
-            },
-            {
-              "value": 0
-            }
-          ]
+          "opacity": {
+            "field": "animateOpacity"
+          }
         }
       }
     },
     {
       "name": "shadow-path",
       "type": "path",
       "from": {
         "data": "routes"
       },
       "encode": {
         "enter": {
           "tooltip": {
             "field": "tooltip_data"
           },
           "strokeWidth": {
             "value": 10
           },
           "strokeOpacity": {
             "value": 0.1
           }
         },
         "update": {
           "path": {
             "field": "path"
           },
-          "stroke": [
-            {
-              "test": "datum._source.color && datum._source.color != ''",
-              "field": "_source.color"
-            },
-            {
-              "value": "#54B399"
-            }
-          ]
+          "stroke": {
+            "field": "color"
+          },
+          "opacity": {
+            "field": "animateOpacity"
+          }
         }
       }
     },
     {
       "name": "moving-arrow",
       "type": "text",
       "from": {
         "data": "animatedRoutes"
       },
       "encode": {
         "enter": {
           "text": {
             "value": "➤"
           },
           "align": {
             "value": "center"
           },
           "baseline": {
             "value": "middle"
           },
           "fontWeight": {
             "value": "bold"
           },
           "fontSize": {
             "value": 11
           },
           "dy": {
             "value": 1
           },
-          "fill": [
-            {
-              "test": "datum._source.color && datum._source.color != ''",
-              "field": "_source.color"
-            },
-            {
-              "value": "#54B399"
-            }
-          ]
+          "fill": {
+            "field": "color"
+          }
         },
         "update": {
           "x": {
             "signal": "pow(1 - datum.fraction, 3) * datum.sourceX + 3 * pow(1 - datum.fraction, 2) * datum.fraction * datum.control1X + 3 * (1 - datum.fraction) * pow(datum.fraction, 2) * datum.control2X + pow(datum.fraction, 3) * datum.targetX"
           },
           "y": {
             "signal": "pow(1 - datum.fraction, 3) * datum.sourceY + 3 * pow(1 - datum.fraction, 2) * datum.fraction * datum.control1Y + 3 * (1 - datum.fraction) * pow(datum.fraction, 2) * datum.control2Y + pow(datum.fraction, 3) * datum.targetY"
           },
           "angle": {
             "signal": "atan2(3 * pow(1 - datum.fraction, 2) * (datum.control1Y - datum.sourceY) + 6 * (1 - datum.fraction) * datum.fraction * (datum.control2Y - datum.control1Y) + 3 * pow(datum.fraction, 2) * (datum.targetY - datum.control2Y), 3 * pow(1 - datum.fraction, 2) * (datum.control1X - datum.sourceX) + 6 * (1 - datum.fraction) * datum.fraction * (datum.control2X - datum.control1X) + 3 * pow(datum.fraction, 2) * (datum.targetX - datum.control2X)) * 180 / PI"
           },
-          "opacity": [
-            {
-              "test": "datum._source.animate && datum._source.animate == true",
-              "value": 1
-            },
-            {
-              "value": 0
-            }
-          ]
+          "opacity": {
+            "field": "animateOpacity"
+          }
         }
       }
     },
     {
       "name": "fixed-arrow",
-      "description": "A fixed arrow at the end of the path",
       "type": "text",
       "from": {
         "data": "routes"
       },
       "encode": {
         "enter": {
           "fontWeight": {
             "value": "bold"
           },
           "fontSize": {
             "value": 11
           },
           "text": {
             "value": "➤"
           },
           "align": {
             "value": "center"
           },
           "baseline": {
             "value": "middle"
           },
-          "fill": [
-            {
-              "test": "datum._source.color && datum._source.color != ''",
-              "field": "_source.color"
-            },
-            {
-              "value": "#54B399"
-            }
-          ],
+          "fill": {
+            "field": "color"
+          },
           "dy": {
             "value": 1
           }
         },
         "update": {
           "x": {
             "field": "targetX"
           },
           "y": {
             "field": "targetY"
           },
           "angle": {
             "signal": "atan2(datum.targetY - datum.control2Y, datum.targetX - datum.control2X) * 180 / PI"
           }
         }
       }
     },
     {
       "name": "source-labels",
-      "description": "The lables to show at the start of the path",
       "type": "text",
       "from": {
         "data": "routes"
       },
       "encode": {
         "enter": {
           "text": {
-            "field": "_source.source_label"
+            "field": "source_label"
           },
           "fill": {
             "value": "#54B399"
           },
           "align": {
             "value": "center"
           },
           "baseline": {
             "value": "middle"
           },
           "fontWeight": {
             "value": "bold"
           },
           "fontSize": {
             "value": 11
           }
         },
         "update": {
           "x": {
             "field": "sourceX"
           },
           "y": {
             "field": "sourceY"
           },
           "opacity": [
             {
               "test": "showLables == true",
               "value": 1
             },
             {
               "value": 0
             }
           ]
         }
       }
     },
     {
       "name": "destination-labels",
       "description": "The lables to show at the end of the path",
       "type": "text",
       "from": {
         "data": "routes"
       },
       "encode": {
         "enter": {
           "text": {
-            "field": "_source.destination_label"
+            "field": "destination_label"
           },
           "fill": {
             "value": "#54B399"
           },
           "fontSize": {
             "value": 11
           },
           "align": {
             "value": "center"
           },
           "baseline": {
             "value": "middle"
           },
           "fontWeight": {
             "value": "bold"
           }
         },
         "update": {
           "x": {
             "field": "targetX"
           },
           "y": {
             "field": "targetY"
           },
           "opacity": [
             {
               "test": "showLables == true",
               "value": 1
             },
             {
               "value": 0
             }
           ]
         }
       }
     },
     {
       "name": "pulsing-circle",
-      "description": "A circle pulsing at the end of the path",
       "type": "arc",
       "from": {
         "data": "animatedRoutes"
       },
       "encode": {
         "enter": {
           "startAngle": {
             "value": 0
           },
           "endAngle": {
             "value": 6.29
           },
           "innerRadius": {
             "value": 0
           },
-          "stroke": [
-            {
-              "test": "datum._source.color && datum._source.color != ''",
-              "field": "_source.color"
-            },
-            {
-              "value": "#54B399"
-            }
-          ],
+          "stroke": {
+            "field": "color"
+          },
           "strokeWidth": {
             "value": 2
           }
         },
         "update": {
           "x": [
             {
-              "test": "datum._source.pulse_at_source && datum._source.pulse_at_source == true",
+              "test": "datum.fields['pulse_at_source'] && datum.fields['pulse_at_source'] == true",
               "field": "sourceX"
             },
             {
               "field": "targetX"
             }
           ],
           "y": [
             {
-              "test": "datum._source.pulse_at_source && datum._source.pulse_at_source == true",
+              "test": "datum.fields['pulse_at_source'] && datum.fields['pulse_at_source'] == true",
               "field": "sourceY"
             },
             {
               "field": "targetY"
             }
           ],
           "outerRadius": {
-            "signal": "radius"
+            "signal": "20 * fraction2"
           },
           "strokeOpacity": {
             "signal": "1 - fraction2"
           },
-          "opacity": [
-            {
-              "test": "datum._source.animate && datum._source.animate == true",
-              "value": 1
-            },
-            {
-              "value": 0
-            }
-          ]
+          "opacity": {
+            "field": "animateOpacity"
+          }
         }
       }
     }
   ]
 }
```

**Comment by navnit-crest** [comment]
> Looks good. Can you please add a screenshot to the PR description? (ideally before and after)

I've added screenshots of the Threat Map dashboard, showing its state before and after the fixes in the PR description.

**Review by chrisberkhout — APPROVED** [review]
Thanks! Looks great.

**Comment by iamhowardtheduck** [comment]
Observed the same "Invalid SVG path, incorrect parameter type" error on version  0.1.1

image
image

**Pull request merged**